go/parquet/internal/encoding/levels.go

```diff
@@ -0,0 +1,284 @@
+	case format.Encoding_BIT_PACKED:
+		l.bit.Clear()
+	default:
+		panic("parquet: unknown encoding type")
+	}
+}
+
+// Init is called to set up the desired encoding type, max level and underlying writer for a
+// level encoder to control where the resulting encoded buffer will end up.
+func (l *LevelEncoder) Init(encoding parquet.Encoding, maxLvl int16, w io.WriterAt) {
+	l.bitWidth = bits.Len64(uint64(maxLvl))
+	l.encoding = format.Encoding(encoding)
+	switch l.encoding {
+	case format.Encoding_RLE:
+		l.rle = utils.NewRleEncoder(w, l.bitWidth)
+	case format.Encoding_BIT_PACKED:
+		l.bit = utils.NewBitWriter(w)
+	default:
+		panic("parquet: unknown encoding type for levels")
+	}
+}
+
+// EncodeNoFlush encodes the provided levels in the encoder, but doesn't flush
+// the buffer and return it yet, appending these encoded values. Returns the number
+// of values encoded.
+func (l *LevelEncoder) EncodeNoFlush(lvls []int16) int {
+	nencoded := 0
+	if l.rle == nil && l.bit == nil {
+		panic("parquet: level encoders are not initialized")
+	}
+
+	switch l.encoding {
+	case format.Encoding_RLE:
+		for _, level := range lvls {
+			if !l.rle.Put(uint64(level)) {
+				break
+			}
+			nencoded++
+		}
+	default:
+		for _, level := range lvls {
+			if l.bit.WriteValue(uint64(level), uint(l.bitWidth)) != nil {
+				break
+			}
+			nencoded++
+		}
+	}
+	return nencoded
+}
+
+// Flush flushes out any encoded data to the underlying writer.
+func (l *LevelEncoder) Flush() {
+	if l.rle == nil && l.bit == nil {
+		panic("parquet: level encoders are not initialized")
+	}
+
+	switch l.encoding {
+	case format.Encoding_RLE:
+		l.rleLen = l.rle.Flush()
+	default:
+		l.bit.Flush(false)
+	}
+}
+
+// Encode encodes the slice of definition or repetition levels based on
+// the currently configured encoding type and returns the number of
+// values that were encoded.
+func (l *LevelEncoder) Encode(lvls []int16) int {
+	nencoded := 0
+	if l.rle == nil && l.bit == nil {
+		panic("parquet: level encoders are not initialized")
+	}
+
+	switch l.encoding {
+	case format.Encoding_RLE:
+		for _, level := range lvls {
+			if !l.rle.Put(uint64(level)) {
+				break
+			}
+			nencoded++
+		}
+		l.rleLen = l.rle.Flush()
+	default:
+		for _, level := range lvls {
+			if l.bit.WriteValue(uint64(level), uint(l.bitWidth)) != nil {
+				break
+			}
+			nencoded++
+		}
+		l.bit.Flush(false)
+	}
+	return nencoded
+}
+
+// Len returns the number of bytes that were written as Run Length encoded
+// levels, this is only valid for run length encoding and will panic if using
+// deprecated bit packed encoding.
+func (l *LevelEncoder) Len() int {
+	if l.encoding != format.Encoding_RLE {
+		panic("parquet: level encoder, only implemented for RLE")
+	}
+	return l.rleLen
+}
+
+// LevelDecoder handles the decoding of repetition and definition levels from a
+// parquet file supporting bit packed and run length encoded values.
+type LevelDecoder struct {
+	bitWidth  int
+	remaining int
```

Comment:
a comment on what remaining is might be useful.

Reply:
done.